src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs

```diff
@@ -0,0 +1,235 @@
+                    // Then repeatedly generate a value in that outer range until we get one within the inner range.
+                    int bits = Log2Ceiling((ulong)maxValue);
+                    while (true)
+                    {
+                        ulong result = NextUInt64() >> (sizeof(ulong) * 8 - bits);
+                        if (result < (ulong)maxValue)
+                        {
+                            return (long)result;
+                        }
+                    }
+                }
+
+                Debug.Assert(maxValue == 0 || maxValue == 1);
+                return 0;
+            }
+
+            public override long NextInt64(long minValue, long maxValue)
+            {
+                ulong exclusiveRange = (ulong)(maxValue - minValue);
+
+                if (exclusiveRange <= int.MaxValue)
+                {
+                    return Next((int)exclusiveRange) + minValue;
+                }
+
+                if (exclusiveRange > 1)
+                {
+                    // Narrow down to the smallest range [0, 2^bits] that contains maxValue.
+                    // Then repeatedly generate a value in that outer range until we get one within the inner range.
+                    int bits = Log2Ceiling(exclusiveRange);
+                    while (true)
+                    {
+                        ulong result = NextUInt64() >> (sizeof(ulong) * 8 - bits);
+                        if (result < exclusiveRange)
+                        {
+                            return (long)result + minValue;
+                        }
+                    }
+                }
+
+                Debug.Assert(minValue == maxValue || minValue + 1 == maxValue);
+                return minValue;
+            }
+
+            public override void NextBytes(byte[] buffer) => NextBytes((Span<byte>)buffer);
+
+            public override unsafe void NextBytes(Span<byte> buffer)
+            {
+                while (buffer.Length >= sizeof(uint))
+                {
+                    Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(buffer), NextUInt32());
+                    buffer = buffer.Slice(sizeof(uint));
+                }
+
+                if (!buffer.IsEmpty)
+                {
+                    uint next = NextUInt32();
+                    byte* remainingBytes = (byte*)&next;
+                    Debug.Assert(buffer.Length < sizeof(uint));
+                    for (int i = 0; i < buffer.Length; i++)
+                    {
+                        buffer[i] = remainingBytes[i];
+                    }
+                }
+            }
+
+            public override double NextDouble() =>
+                // See comment in Xoshiro256StarStarImpl. But creating 64 random bits is expensive compared to the
+                // legacy implementation that only created 31, so for now we stick with the same 31 bits.
+                (NextUInt32() >> 1) * (1.0 / (1u << 31));
```

Comment:
@colgreen, @tannergooding, any concerns with this?  I initially used NextUInt64 (which just calls NextUInt32 twice, casts, and shift), but it showed around a 60% regression, and given the purpose of this type and the quality of the current prng being replaced, this seemed an appropriate compromise.

Reply:
So, just to confirm. The existing `new Random(seed).NextDouble()` only returns 31-bits?
Is creating 64 random bits only expensive on 32-bit?

Reply:
> The existing new Random(seed).NextDouble() only returns 31-bits?

Yes.
```C#
public virtual double NextDouble() => Sample();

protected virtual double Sample() =>
    InternalSample() * (1.0 / int.MaxValue);

private int InternalSample() { ... }
```

> Is creating 64 random bits only expensive on 32-bit?

Yes. (And, to be clear, I'm talking relatively... everything here is measured in a few nanoseconds. But, for example, on 32-bit it was regressing from around 9ns to around 14ns.)

Reply:
In that case, I'd suggest this at least be:
```csharp
if (Environment.Is64BitProcess)
{
    return (NextUInt64() >> 11) * (1.0 / (1UL << 53));
}
else
{
    return (NextUInt32() >> 1) * (1.0 / (1U << 31));
}
```

I think it might also be worth considering taking the regression for 32-bit given the increased distribution and therefore usability of `double` for other scenarios, but I don't have particularly strong feelings about it.

Reply:
We do not care as much about performance on 32-bit platforms. I think it would be ok to regress 32-bit platforms to keep the code simpler.

Reply:
> In that case, I'd suggest this at least be:

In a 64-bit process, this implementation won't be used at all.

> We do not care as much about performance on 32-bit platforms. I think it would be ok to regress 32-bit platforms to keep the code simpler.

@jkotas, is your comment specific to @tannergooding's?  Or are you suggesting I delete the Xoshiro128StarStar implementation (which is used now for 32-bit) and just always use the Xoshiro256StarStar?

Reply:
> In a 64-bit process, this implementation won't be used at all.

Ah, I see. I missed that part.

Reply:
(Or I guess alternatively just continue to use the existing implementation on 32-bit and only use the new one for 64-bit?)

Reply:
I think it would be fine to use `Xoshiro256StarStar` everywhere to keep things simple and take the hit on 32-bit platforms.

This is not crypto, so we have some freedom here, but having lower quality generator for 32-bit sounds a bit suspect to me. I do not know about any other platform with implementation like that.

Reply:
I do not have strong opinion about this.

Reply:
Ok, I'll change the Xoshiro128StarStar.NextDouble implementation to use 53 instead of 31 bits and call it a day.

Reply:
Late to the party, but I agree with Jan. 

A different PRNG underneath to avoid the performance regression made sense and was worth exploring, but reverting to the old way of using the PRNG bits for generating floats seems like a step too far in the balance between different goals.

So overall, keeping it simple and keeping the implementations the same for 32 and 64 bit makes sense to me. Yes there's a perf regression, but (a) this new code fixes all of the outstanding issues raised (statistical quirks and such), (b) is still actually a lot faster in in some scenarios (the constructor, and NextBytes), and (c) is still very fast even with the regression on 32 bit. So think this is the sweet spot.

Reply:
I have a lot of trouble swallowing a 3x regression for random.Next(int) and random.Next(int, int), which from what I've seen are two of the most common operations performed on a Random.  I reverted Xoshiro128StarStar.NextDouble to use 53 bits as was suggested, but I don't think we should use Xoshiro256StarStar on 32-bit.

Reply:
ahh just saw you've kept the Xoshiro128StarStar impl for 32 bit. I think that's fine. There's a slight niggle here in that the low bits of these PRNGs fail some statistical tests, but probably not in an important way. I just worry a bit about using 2 x 32 bit values to get the 53 bits for NextDouble(), as those lower quality random bits are now somewhere in the middle of the bits, instead of right at the low significance end.  They are still 33+ bits along though, so overall I'm not concerned - especially since the old way used only 31 bits of randomness.

It's perhaps a slight quirk that the 32 bit world has a slightly different statistical quality of PRNG to 64 bit, but at this point I'd say it's a coin flip decision. Xoshiro128StarStar does have a nice speedup in 32 bit (avoids the perf regression), so I think I am all good with this.

Reply:
Random thought (heh) - are we going to implement NextHalf() ?

Reply:
> Random thought (heh) - are we going to implement NextHalf() ?

I don't see a proposal for it.
cc: @tannergooding, @pgovind